<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>ObservableRepeatTest.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">RxJava_2_x$Unnamed.exec</a> &gt; <a href="index.source.html" class="el_package">io.reactivex.internal.operators.observable</a> &gt; <span class="el_source">ObservableRepeatTest.java</span></div><h1>ObservableRepeatTest.java</h1><pre class="source lang-java linenums">/**
 * Copyright (c) 2016-present, RxJava Contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the &quot;License&quot;); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an &quot;AS IS&quot; BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
 * the License for the specific language governing permissions and limitations under the License.
 */

package io.reactivex.internal.operators.observable;

import static org.junit.Assert.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

import java.util.*;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import org.junit.Test;

import io.reactivex.*;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.disposables.*;
import io.reactivex.exceptions.TestException;
import io.reactivex.functions.*;
import io.reactivex.observers.TestObserver;
import io.reactivex.plugins.RxJavaPlugins;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subjects.PublishSubject;

<span class="fc" id="L37">public class ObservableRepeatTest {</span>

    @Test(timeout = 2000)
    public void testRepetition() {
<span class="fc" id="L41">        int num = 10;</span>
<span class="fc" id="L42">        final AtomicInteger count = new AtomicInteger();</span>
<span class="fc" id="L43">        int value = Observable.unsafeCreate(new ObservableSource&lt;Integer&gt;() {</span>

            @Override
            public void subscribe(final Observer&lt;? super Integer&gt; o) {
<span class="fc" id="L47">                o.onNext(count.incrementAndGet());</span>
<span class="fc" id="L48">                o.onComplete();</span>
<span class="fc" id="L49">            }</span>
<span class="fc" id="L50">        }).repeat().subscribeOn(Schedulers.computation())</span>
<span class="fc" id="L51">        .take(num).blockingLast();</span>

<span class="fc" id="L53">        assertEquals(num, value);</span>
<span class="fc" id="L54">    }</span>

    @Test(timeout = 2000)
    public void testRepeatTake() {
<span class="fc" id="L58">        Observable&lt;Integer&gt; xs = Observable.just(1, 2);</span>
<span class="fc" id="L59">        Object[] ys = xs.repeat().subscribeOn(Schedulers.newThread()).take(4).toList().blockingGet().toArray();</span>
<span class="fc" id="L60">        assertArrayEquals(new Object[] { 1, 2, 1, 2 }, ys);</span>
<span class="fc" id="L61">    }</span>

    @Test(timeout = 20000)
    public void testNoStackOverFlow() {
<span class="fc" id="L65">        Observable.just(1).repeat().subscribeOn(Schedulers.newThread()).take(100000).blockingLast();</span>
<span class="fc" id="L66">    }</span>

    @Test
    public void testRepeatTakeWithSubscribeOn() throws InterruptedException {

<span class="fc" id="L71">        final AtomicInteger counter = new AtomicInteger();</span>
<span class="fc" id="L72">        Observable&lt;Integer&gt; oi = Observable.unsafeCreate(new ObservableSource&lt;Integer&gt;() {</span>

            @Override
            public void subscribe(Observer&lt;? super Integer&gt; sub) {
<span class="fc" id="L76">                sub.onSubscribe(Disposables.empty());</span>
<span class="fc" id="L77">                counter.incrementAndGet();</span>
<span class="fc" id="L78">                sub.onNext(1);</span>
<span class="fc" id="L79">                sub.onNext(2);</span>
<span class="fc" id="L80">                sub.onComplete();</span>
<span class="fc" id="L81">            }</span>
<span class="fc" id="L82">        }).subscribeOn(Schedulers.newThread());</span>

<span class="fc" id="L84">        Object[] ys = oi.repeat().subscribeOn(Schedulers.newThread()).map(new Function&lt;Integer, Integer&gt;() {</span>

            @Override
            public Integer apply(Integer t1) {
                try {
<span class="fc" id="L89">                    Thread.sleep(50);</span>
<span class="nc" id="L90">                } catch (InterruptedException e) {</span>
<span class="nc" id="L91">                    e.printStackTrace();</span>
<span class="fc" id="L92">                }</span>
<span class="fc" id="L93">                return t1;</span>
            }

<span class="fc" id="L96">        }).take(4).toList().blockingGet().toArray();</span>

<span class="fc" id="L98">        assertEquals(2, counter.get());</span>
<span class="fc" id="L99">        assertArrayEquals(new Object[] { 1, 2, 1, 2 }, ys);</span>
<span class="fc" id="L100">    }</span>

    @Test(timeout = 2000)
    public void testRepeatAndTake() {
<span class="fc" id="L104">        Observer&lt;Object&gt; o = TestHelper.mockObserver();</span>

<span class="fc" id="L106">        Observable.just(1).repeat().take(10).subscribe(o);</span>

<span class="fc" id="L108">        verify(o, times(10)).onNext(1);</span>
<span class="fc" id="L109">        verify(o).onComplete();</span>
<span class="fc" id="L110">        verify(o, never()).onError(any(Throwable.class));</span>
<span class="fc" id="L111">    }</span>

    @Test(timeout = 2000)
    public void testRepeatLimited() {
<span class="fc" id="L115">        Observer&lt;Object&gt; o = TestHelper.mockObserver();</span>

<span class="fc" id="L117">        Observable.just(1).repeat(10).subscribe(o);</span>

<span class="fc" id="L119">        verify(o, times(10)).onNext(1);</span>
<span class="fc" id="L120">        verify(o).onComplete();</span>
<span class="fc" id="L121">        verify(o, never()).onError(any(Throwable.class));</span>
<span class="fc" id="L122">    }</span>

    @Test(timeout = 2000)
    public void testRepeatError() {
<span class="fc" id="L126">        Observer&lt;Object&gt; o = TestHelper.mockObserver();</span>

<span class="fc" id="L128">        Observable.error(new TestException()).repeat(10).subscribe(o);</span>

<span class="fc" id="L130">        verify(o).onError(any(TestException.class));</span>
<span class="fc" id="L131">        verify(o, never()).onNext(any());</span>
<span class="fc" id="L132">        verify(o, never()).onComplete();</span>

<span class="fc" id="L134">    }</span>

    @Test(timeout = 2000)
    public void testRepeatZero() {
<span class="fc" id="L138">        Observer&lt;Object&gt; o = TestHelper.mockObserver();</span>

<span class="fc" id="L140">        Observable.just(1).repeat(0).subscribe(o);</span>

<span class="fc" id="L142">        verify(o).onComplete();</span>
<span class="fc" id="L143">        verify(o, never()).onNext(any());</span>
<span class="fc" id="L144">        verify(o, never()).onError(any(Throwable.class));</span>
<span class="fc" id="L145">    }</span>

    @Test(timeout = 2000)
    public void testRepeatOne() {
<span class="fc" id="L149">        Observer&lt;Object&gt; o = TestHelper.mockObserver();</span>

<span class="fc" id="L151">        Observable.just(1).repeat(1).subscribe(o);</span>

<span class="fc" id="L153">        verify(o).onComplete();</span>
<span class="fc" id="L154">        verify(o, times(1)).onNext(any());</span>
<span class="fc" id="L155">        verify(o, never()).onError(any(Throwable.class));</span>
<span class="fc" id="L156">    }</span>

    /** Issue #2587. */
    @Test
    public void testRepeatAndDistinctUnbounded() {
<span class="fc" id="L161">        Observable&lt;Integer&gt; src = Observable.fromIterable(Arrays.asList(1, 2, 3, 4, 5))</span>
<span class="fc" id="L162">                .take(3)</span>
<span class="fc" id="L163">                .repeat(3)</span>
<span class="fc" id="L164">                .distinct();</span>

<span class="fc" id="L166">        TestObserver&lt;Integer&gt; to = new TestObserver&lt;Integer&gt;();</span>

<span class="fc" id="L168">        src.subscribe(to);</span>

<span class="fc" id="L170">        to.assertNoErrors();</span>
<span class="fc" id="L171">        to.assertTerminated();</span>
<span class="fc" id="L172">        to.assertValues(1, 2, 3);</span>
<span class="fc" id="L173">    }</span>

    /** Issue #2844: wrong target of request. */
    @Test(timeout = 3000)
    public void testRepeatRetarget() {
<span class="fc" id="L178">        final List&lt;Integer&gt; concatBase = new ArrayList&lt;Integer&gt;();</span>
<span class="fc" id="L179">        TestObserver&lt;Integer&gt; to = new TestObserver&lt;Integer&gt;();</span>
<span class="fc" id="L180">        Observable.just(1, 2)</span>
<span class="fc" id="L181">        .repeat(5)</span>
<span class="fc" id="L182">        .concatMap(new Function&lt;Integer, Observable&lt;Integer&gt;&gt;() {</span>
            @Override
            public Observable&lt;Integer&gt; apply(Integer x) {
<span class="fc" id="L185">                System.out.println(&quot;testRepeatRetarget -&gt; &quot; + x);</span>
<span class="fc" id="L186">                concatBase.add(x);</span>
<span class="fc" id="L187">                return Observable.&lt;Integer&gt;empty()</span>
<span class="fc" id="L188">                        .delay(200, TimeUnit.MILLISECONDS);</span>
            }
        })
<span class="fc" id="L191">        .subscribe(to);</span>

<span class="fc" id="L193">        to.awaitTerminalEvent();</span>
<span class="fc" id="L194">        to.assertNoErrors();</span>
<span class="fc" id="L195">        to.assertNoValues();</span>

<span class="fc" id="L197">        assertEquals(Arrays.asList(1, 2, 1, 2, 1, 2, 1, 2, 1, 2), concatBase);</span>
<span class="fc" id="L198">    }</span>

    @Test
    public void repeatUntil() {
<span class="fc" id="L202">        Observable.just(1)</span>
<span class="fc" id="L203">        .repeatUntil(new BooleanSupplier() {</span>
            @Override
            public boolean getAsBoolean() throws Exception {
<span class="fc" id="L206">                return false;</span>
            }
        })
<span class="fc" id="L209">        .take(5)</span>
<span class="fc" id="L210">        .test()</span>
<span class="fc" id="L211">        .assertResult(1, 1, 1, 1, 1);</span>
<span class="fc" id="L212">    }</span>

    @Test
    public void repeatLongPredicateInvalid() {
        try {
<span class="nc" id="L217">            Observable.just(1).repeat(-99);</span>
<span class="nc" id="L218">            fail(&quot;Should have thrown&quot;);</span>
<span class="fc" id="L219">        } catch (IllegalArgumentException ex) {</span>
<span class="fc" id="L220">            assertEquals(&quot;times &gt;= 0 required but it was -99&quot;, ex.getMessage());</span>
<span class="nc" id="L221">        }</span>
<span class="fc" id="L222">    }</span>

    @Test
    public void repeatUntilError() {
<span class="fc" id="L226">        Observable.error(new TestException())</span>
<span class="fc" id="L227">        .repeatUntil(new BooleanSupplier() {</span>
            @Override
            public boolean getAsBoolean() throws Exception {
<span class="nc" id="L230">                return true;</span>
            }
        })
<span class="fc" id="L233">        .test()</span>
<span class="fc" id="L234">        .assertFailure(TestException.class);</span>
<span class="fc" id="L235">    }</span>

    @Test
    public void repeatUntilFalse() {
<span class="fc" id="L239">        Observable.just(1)</span>
<span class="fc" id="L240">        .repeatUntil(new BooleanSupplier() {</span>
            @Override
            public boolean getAsBoolean() throws Exception {
<span class="fc" id="L243">                return true;</span>
            }
        })
<span class="fc" id="L246">        .test()</span>
<span class="fc" id="L247">        .assertResult(1);</span>
<span class="fc" id="L248">    }</span>

    @Test
    public void repeatUntilSupplierCrash() {
<span class="fc" id="L252">        Observable.just(1)</span>
<span class="fc" id="L253">        .repeatUntil(new BooleanSupplier() {</span>
            @Override
            public boolean getAsBoolean() throws Exception {
<span class="fc" id="L256">                throw new TestException();</span>
            }
        })
<span class="fc" id="L259">        .test()</span>
<span class="fc" id="L260">        .assertFailure(TestException.class, 1);</span>
<span class="fc" id="L261">    }</span>

    @Test
    public void shouldDisposeInnerObservable() {
<span class="fc" id="L265">      final PublishSubject&lt;Object&gt; subject = PublishSubject.create();</span>
<span class="fc" id="L266">      final Disposable disposable = Observable.just(&quot;Leak&quot;)</span>
<span class="fc" id="L267">          .repeatWhen(new Function&lt;Observable&lt;Object&gt;, ObservableSource&lt;Object&gt;&gt;() {</span>
            @Override
            public ObservableSource&lt;Object&gt; apply(Observable&lt;Object&gt; completions) throws Exception {
<span class="fc" id="L270">                return completions.switchMap(new Function&lt;Object, ObservableSource&lt;Object&gt;&gt;() {</span>
                    @Override
                    public ObservableSource&lt;Object&gt; apply(Object ignore) throws Exception {
<span class="fc" id="L273">                        return subject;</span>
                    }
                });
            }
        })
<span class="fc" id="L278">          .subscribe();</span>

<span class="fc" id="L280">      assertTrue(subject.hasObservers());</span>
<span class="fc" id="L281">      disposable.dispose();</span>
<span class="fc" id="L282">      assertFalse(subject.hasObservers());</span>
<span class="fc" id="L283">    }</span>

    @Test
    public void testRepeatWhen() {
<span class="fc" id="L287">        Observable.error(new TestException())</span>
<span class="fc" id="L288">        .repeatWhen(new Function&lt;Observable&lt;Object&gt;, ObservableSource&lt;Object&gt;&gt;() {</span>
            @Override
            public ObservableSource&lt;Object&gt; apply(Observable&lt;Object&gt; v) throws Exception {
<span class="fc" id="L291">                return v.delay(10, TimeUnit.SECONDS);</span>
            }
        })
<span class="fc" id="L294">        .test()</span>
<span class="fc" id="L295">        .awaitDone(5, TimeUnit.SECONDS)</span>
<span class="fc" id="L296">        .assertFailure(TestException.class);</span>
<span class="fc" id="L297">    }</span>

    @Test
    public void whenTake() {
<span class="fc" id="L301">        Observable.range(1, 3).repeatWhen(new Function&lt;Observable&lt;Object&gt;, ObservableSource&lt;Object&gt;&gt;() {</span>
            @Override
            public ObservableSource&lt;Object&gt; apply(Observable&lt;Object&gt; handler) throws Exception {
<span class="fc" id="L304">                return handler.take(2);</span>
            }
        })
<span class="fc" id="L307">        .test()</span>
<span class="fc" id="L308">        .assertResult(1, 2, 3, 1, 2, 3);</span>
<span class="fc" id="L309">    }</span>

    @Test
    public void handlerError() {
<span class="fc" id="L313">        Observable.range(1, 3)</span>
<span class="fc" id="L314">        .repeatWhen(new Function&lt;Observable&lt;Object&gt;, ObservableSource&lt;Object&gt;&gt;() {</span>
            @Override
            public ObservableSource&lt;Object&gt; apply(Observable&lt;Object&gt; v) throws Exception {
<span class="fc" id="L317">                return v.map(new Function&lt;Object, Object&gt;() {</span>
                    @Override
                    public Object apply(Object w) throws Exception {
<span class="fc" id="L320">                        throw new TestException();</span>
                    }
                });
            }
        })
<span class="fc" id="L325">        .test()</span>
<span class="fc" id="L326">        .assertFailure(TestException.class, 1, 2, 3);</span>
<span class="fc" id="L327">    }</span>

    @Test
    public void noCancelPreviousRepeat() {
<span class="fc" id="L331">        final AtomicInteger counter = new AtomicInteger();</span>

<span class="fc" id="L333">        Observable&lt;Integer&gt; source = Observable.just(1).doOnDispose(new Action() {</span>
            @Override
            public void run() throws Exception {
<span class="nc" id="L336">                counter.getAndIncrement();</span>
<span class="nc" id="L337">            }</span>
        });

<span class="fc" id="L340">        source.repeat(5)</span>
<span class="fc" id="L341">        .test()</span>
<span class="fc" id="L342">        .assertResult(1, 1, 1, 1, 1);</span>

<span class="fc" id="L344">        assertEquals(0, counter.get());</span>
<span class="fc" id="L345">    }</span>

    @Test
    public void noCancelPreviousRepeatUntil() {
<span class="fc" id="L349">        final AtomicInteger counter = new AtomicInteger();</span>

<span class="fc" id="L351">        Observable&lt;Integer&gt; source = Observable.just(1).doOnDispose(new Action() {</span>
            @Override
            public void run() throws Exception {
<span class="nc" id="L354">                counter.getAndIncrement();</span>
<span class="nc" id="L355">            }</span>
        });

<span class="fc" id="L358">        final AtomicInteger times = new AtomicInteger();</span>

<span class="fc" id="L360">        source.repeatUntil(new BooleanSupplier() {</span>
            @Override
            public boolean getAsBoolean() throws Exception {
<span class="fc bfc" id="L363" title="All 2 branches covered.">                return times.getAndIncrement() == 4;</span>
            }
        })
<span class="fc" id="L366">        .test()</span>
<span class="fc" id="L367">        .assertResult(1, 1, 1, 1, 1);</span>

<span class="fc" id="L369">        assertEquals(0, counter.get());</span>
<span class="fc" id="L370">    }</span>

    @Test
    public void noCancelPreviousRepeatWhen() {
<span class="fc" id="L374">        final AtomicInteger counter = new AtomicInteger();</span>

<span class="fc" id="L376">        Observable&lt;Integer&gt; source = Observable.just(1).doOnDispose(new Action() {</span>
            @Override
            public void run() throws Exception {
<span class="nc" id="L379">                counter.getAndIncrement();</span>
<span class="nc" id="L380">            }</span>
        });

<span class="fc" id="L383">        final AtomicInteger times = new AtomicInteger();</span>

<span class="fc" id="L385">        source.repeatWhen(new Function&lt;Observable&lt;Object&gt;, ObservableSource&lt;?&gt;&gt;() {</span>
            @Override
            public ObservableSource&lt;?&gt; apply(Observable&lt;Object&gt; e) throws Exception {
<span class="fc" id="L388">                return e.takeWhile(new Predicate&lt;Object&gt;() {</span>
                    @Override
                    public boolean test(Object v) throws Exception {
<span class="fc bfc" id="L391" title="All 2 branches covered.">                        return times.getAndIncrement() &lt; 4;</span>
                    }
                });
            }
        })
<span class="fc" id="L396">        .test()</span>
<span class="fc" id="L397">        .assertResult(1, 1, 1, 1, 1);</span>

<span class="fc" id="L399">        assertEquals(0, counter.get());</span>
<span class="fc" id="L400">    }</span>

    @Test
    public void repeatFloodNoSubscriptionError() {
<span class="fc" id="L404">        List&lt;Throwable&gt; errors = TestHelper.trackPluginErrors();</span>

        try {
<span class="fc" id="L407">            final PublishSubject&lt;Integer&gt; source = PublishSubject.create();</span>
<span class="fc" id="L408">            final PublishSubject&lt;Integer&gt; signaller = PublishSubject.create();</span>

<span class="fc bfc" id="L410" title="All 2 branches covered.">            for (int i = 0; i &lt; TestHelper.RACE_DEFAULT_LOOPS; i++) {</span>

<span class="fc" id="L412">                TestObserver&lt;Integer&gt; to = source.take(1)</span>
<span class="fc" id="L413">                .repeatWhen(new Function&lt;Observable&lt;Object&gt;, ObservableSource&lt;Integer&gt;&gt;() {</span>
                    @Override
                    public ObservableSource&lt;Integer&gt; apply(Observable&lt;Object&gt; v)
                            throws Exception {
<span class="fc" id="L417">                        return signaller;</span>
                    }
<span class="fc" id="L419">                }).test();</span>

<span class="fc" id="L421">                Runnable r1 = new Runnable() {</span>
                    @Override
                    public void run() {
<span class="fc bfc" id="L424" title="All 2 branches covered.">                        for (int i = 0; i &lt; TestHelper.RACE_DEFAULT_LOOPS; i++) {</span>
<span class="fc" id="L425">                            source.onNext(1);</span>
                        }
<span class="fc" id="L427">                    }</span>
                };
<span class="fc" id="L429">                Runnable r2 = new Runnable() {</span>
                    @Override
                    public void run() {
<span class="fc bfc" id="L432" title="All 2 branches covered.">                        for (int i = 0; i &lt; TestHelper.RACE_DEFAULT_LOOPS; i++) {</span>
<span class="fc" id="L433">                            signaller.onNext(1);</span>
                        }
<span class="fc" id="L435">                    }</span>
                };

<span class="fc" id="L438">                TestHelper.race(r1, r2);</span>

<span class="fc" id="L440">                to.dispose();</span>
            }

<span class="pc bpc" id="L443" title="1 of 2 branches missed.">            if (!errors.isEmpty()) {</span>
<span class="nc bnc" id="L444" title="All 2 branches missed.">                for (Throwable e : errors) {</span>
<span class="nc" id="L445">                    e.printStackTrace();</span>
<span class="nc" id="L446">                }</span>
<span class="nc" id="L447">                fail(errors + &quot;&quot;);</span>
            }
        } finally {
<span class="fc" id="L450">            RxJavaPlugins.reset();</span>
        }
<span class="fc" id="L452">    }</span>
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.2.201808211720</span></div></body></html>